api/command.go

```diff
@@ -17,630 +16,386 @@ import (
+		for _, cmd := range teamCmds {
+			if cmd.AutoComplete && !seen[cmd.Id] {
+				cmd.Sanatize()
+				seen[cmd.Trigger] = true
+				commands = append(commands, cmd)
 			}
 		}
-
-		if delay > 10000 {
-			c.Err = model.NewLocAppError("echoCommand", "api.command.echo_command.under.app_error", nil, "")
-			return false
-		}
-
-		if echoSem == nil {
-			// We want one additional thread allowed so we never reach channel lockup
-			echoSem = make(chan bool, maxThreads+1)
-		}
-
-		if len(echoSem) >= maxThreads {
-			c.Err = model.NewLocAppError("echoCommand", "api.command.echo_command.high_volume.app_error", nil, "")
-			return false
-		}
-
-		echoSem <- true
-		go func() {
-			defer func() { <-echoSem }()
-			post := &model.Post{}
-			post.ChannelId = command.ChannelId
-			post.Message = message
-
-			time.Sleep(time.Duration(delay) * time.Second)
-
-			if _, err := CreatePost(c, post, true); err != nil {
-				l4g.Error(utils.T("api.command.echo_command.create.error"), err)
-			}
-		}()
-
-		command.Response = model.RESP_EXECUTED
-		return true
-
-	} else if strings.Index(cmd, command.Command) == 0 {
-		command.AddSuggestion(&model.SuggestCommand{Suggestion: cmd, Description: c.T("api.command.echo_command.description")})
 	}
 
-	return false
+	w.Write([]byte(model.CommandListToJson(commands)))
 }
 
-func meCommand(c *Context, command *model.Command) bool {
-	cmd := cmds["meCommand"]
+func executeCommand(c *Context, w http.ResponseWriter, r *http.Request) {
+	props := model.MapFromJson(r.Body)
+	command := strings.TrimSpace(props["command"])
+	channelId := strings.TrimSpace(props["channelId"])
 
-	if !command.Suggest && strings.Index(command.Command, cmd) == 0 {
-		message := ""
+	if len(command) <= 1 || strings.Index(command, "/") != 0 {
+		c.Err = model.NewLocAppError("executeCommand", "api.command.execute_command.start.app_error", nil, "")
+		return
+	}
 
-		parameters := strings.SplitN(command.Command, " ", 2)
-		if len(parameters) > 1 {
-			message += "*" + parameters[1] + "*"
-		}
+	if len(channelId) > 0 {
+		cchan := Srv.Store.Channel().CheckPermissionsTo(c.Session.TeamId, channelId, c.Session.UserId)
 
-		post := &model.Post{}
-		post.Message = message
-		post.ChannelId = command.ChannelId
-		if _, err := CreatePost(c, post, false); err != nil {
-			l4g.Error(utils.T("api.command.me_command.create.error"), err)
-			return false
+		if !c.HasPermissionsToChannel(cchan, "checkCommand") {
+			return
 		}
-		command.Response = model.RESP_EXECUTED
-		return true
-
-	} else if strings.Index(cmd, command.Command) == 0 {
-		command.AddSuggestion(&model.SuggestCommand{Suggestion: cmd, Description: c.T("api.command.me_command.description")})
 	}
 
-	return false
-}
-
-func shrugCommand(c *Context, command *model.Command) bool {
-	cmd := cmds["shrugCommand"]
+	parts := strings.Split(command, " ")
+	trigger := parts[0][1:]
+	message := strings.Join(parts[1:], " ")
+	provider := GetCommandProvidersProvider(trigger)
```

Comment:
That function name seems confusing, why not just `GetCommandProvider`, `GetCommandProviderFromName` or `GetCommandProviderByTrigger`?